examples/developer-tools/src/output-generators/factory_base.ts

```diff
@@ -123,3 +127,75 @@ jsonDefinitionGenerator.forBlock['factory_base'] = function (
 };
 
 jsonDefinitionGenerator.forBlock['text'] = javascriptGenerator.forBlock['text'];
+
+javascriptDefinitionGenerator.forBlock['factory_base'] = function (
+  block: Blockly.Block,
+  generator: JavascriptDefinitionGenerator,
+) {
+  // TODO: Get a JavaScript-legal name for the block
+  const blockName = block.getFieldValue('NAME');
+  const inputsValue = generator.statementToCode(block, 'INPUTS');
+  const inputs = inputsValue
+    ? generator.prefixLines(inputsValue, generator.INDENT) + '\n'
+    : '';
+
+  const alignInputsValue = block.getFieldValue('INLINE');
+  let alignInputs = '';
+  if (alignInputsValue === 'EXT') {
+    alignInputs = 'this.setInputsInline(false)';
+  } else if (alignInputsValue === 'INT') {
+    alignInputs = 'this.setInputsInline(true)';
+  }
+
+  const createConnectionChecks = function (
+    inputName: string,
+    connectionName: string,
+  ): string {
+    if (block.getInput(inputName)) {
+      const check = generator.valueToCode(
+        block,
+        inputName,
+        JsOrder.FUNCTION_CALL,
+      );
+      return `this.set${connectionName}(true, ${check});`;
+    }
+    return '';
+  };
+  const connectionsList = [];
+  connectionsList.push(createConnectionChecks('OUTPUTCHECK', 'Output'));
+  connectionsList.push(createConnectionChecks('TOPCHECK', 'PreviousStatement'));
+  connectionsList.push(createConnectionChecks('BOTTOMCHECK', 'NextStatement'));
+  const connections = connectionsList
+    .filter((value) => value !== '')
+    .join('\n');
+
+  const tooltip = `this.setTooltip(${
+    generator.valueToCode(block, 'TOOLTIP', JsOrder.ATOMIC) || "''"
+  });`;
+  const helpUrl = `this.setHelpUrl(${
+    generator.valueToCode(block, 'HELPURL', JsOrder.ATOMIC) || "''"
+  });`;
+
+  const colourValue = generator.valueToCode(block, 'COLOUR', JsOrder.ATOMIC);
+  const colour = colourValue !== '' ? `this.setColour(${colourValue});` : '';
+
+  // Filter out empty code string pieces, join them all with newlines, and indent them twice.
+  // Inputs code is already indented once automatically by the generator
+  // since it's in a statement input, so it's not included here.
+  const codeStringPieces = generator.prefixLines(
+    [alignInputs, connections, tooltip, helpUrl, colour]
+      .filter((value) => value !== '')
+      .join('\n'),
+    generator.INDENT + generator.INDENT,
+  );
+  const code = `const ${blockName} = {
+  init: function() {
+${inputs}${codeStringPieces}
+  }
+};
+Blockly.common.defineBlocks({${blockName}: ${blockName}});`;
```

Comment:
This can be shorthanded:
```suggestion
Blockly.common.defineBlocks({${blockName}});`;
```
…though it makes me wonder why we don't just do away with `const ${blockName}` and pass a literal directly to `defineBlocks`.

Reply:
I don't want to just pass to `defineBlocks` directly because I want to make it easy for people to grab the parts of the definition they need. Maybe they want to just create the block definition and export it, and then call `defineBlocks` in another file. I'm optimizing for copy/paste and modifiability, to an extent.

That's also why I didn't shorthand it, because for example in several places we use the actual block name with underscores, but name the block definition object a name that complies with our lint rules, so you end up with `defineBlocks('block_name': blockName)`. I again wanted to make this easy, even though this one is an easier fix than un-inlining the definition.